### ISSUE TITLE
Linker failed with -ObjC linker flag, due to libfreetype on iOS

### ISSUE THREAD

**Issue by jgranick** [openfl/openfl-native#27]
#27 Issue by __louisbl__,

When using the -ObjC linker flag (needed for the Facebook SDK for example), the Linker failed with this error:

```
archive member 'libfreetype.iphonesim.a' with length 594288 is not mach-o or llvm bitcode for architecture i386
```

It seems to be an old nme problem: http://nme-mailing-list.2304833.n4.nabble.com/NME-NME-IOS-failing-to-compile-with-ObjC-flag-in-linker-tp4642103p4642141.html

Rebuilding libnme.a without freetype works.

**Comment by jgranick** [comment]
Comment by __shoebox__:

Any news on this ?

It's quite blocking and forces us to do a fallback on NME 3.5.5...

Thanks !!!

**Comment by jgranick** [comment]
Comment by __shoebox__:

I've discovered than this bug is linked to hxcpp

I have the same bug if i use hxcpp 3.0 + nme 3.5.5

**Comment by jgranick** [comment]
Comment by __puelogames__:

I Just changed haxenme with haxe2 to openfl with haxe3, and now ios admob is with this error =/

**Comment by jgranick** [comment]
Comment by __joshua-jandyco__:

This seems more like a work around but I found that using 

```
-force_load $SRCROOT/$PRODUCT_NAME/lib/$ARCHS/libYourLibThatRequiresObjC.a
```

Instead of -ObjC works in the ios linker-flags works.

**Comment by jgranick** [comment]
Comment by __jgranick__:

Does "-force_load" seem like a good approach?

libfreetype is compiled for i386 when targeting the simulator (you can use otool to see) so I think this must be a bug in the linker? Curious how "-ObjC" would break otherwise valid code

**Comment by jgranick** [comment]
Comment by __baldodavi__:

Facebook support is mandatory for a new product want to be part of multiplatform game environment. I'm not able to help to solve this issue, my level of knowledge in OpenFL platform is on higher level and I need more documentantion to go deep (it could be a limit of mine). I'm working to port some extension, including HypFacebook, to OpenFL and appling results on Stencyl. Work around suggested by joshua-jandyco seems have no effects in my building (I really do not know if I'm applying it in right way).
Is there another way I can try to bypass this issue?
Thanks in advance to all and thanks to all contributors of this great community.
David.

**Comment by jgranick** [comment]
Comment by __joshua-jandyco__:

@baldodavi Feel free to contact me or post what you have done. This is what I did for one of my projects:

```
<dependency name="ndll/iPhone/libAdColony.a" />
<ios linker-flags="-v -fobjc-arc -force_load $SRCROOT/$PRODUCT_NAME/lib/$ARCHS/libAdColony.a" />
```

**Comment by jgranick** [comment]
Comment by __baldodavi__:

Thanks joshua solved using your suggestion.
David.

**Comment by jgranick** [comment]
Comment by __joshua-jandyco__:

@baldodavi Great!  Keep in mind right now you can only have one linker-flags tag in your file. 

I have a PR out that attempts to fix this but @jgranick wants to take linker flags in a more generic direction for next release.

See openfl/hxtools#40

**Comment by jgranick** [comment]
Comment by __madrazo__:

@joshua-jandyco thanks, I had to link -ObjC abd this was very useful.

**Comment by jgranick** [comment]
Comment by __skylerparr__:

I've added the -force_load -force_load $SRCROOT/$PRODUCT_NAME/lib/$ARCHS/libfreetype.iphoneos-v7.a and I still get:
archive member 'libfreetype.iphoneos-v7.a' with length 507640 is not mach-o or llvm bitcode for architecture armv7

Why is this still an issue?  What is the root cause?

**Comment by jgranick** [comment]
Comment by __joshua-jandyco__:

@skylerparr You don't want to force load freetype, you force load whatever library is conflicting and needs objc.

As mentioned in the context of the adcolony library in my comment above.

**Comment by jgranick** [comment]
Comment by __skylerparr__:

I've tried this too:

```
               <dependency name="ndll/iPhone/libhypfacebook.iphoneos-v7.a" />
               <ios linker-flags="-ObjC -force_load$SRCROOT/$PRODUCT_NAME/lib/$ARCHS/libhypfacebook.iphoneos-v7.a" />
```

Same result.

**Comment by jgranick** [comment]
Comment by __SavedByZero__:

Ditto for these "force_load" solutions not working for me.   My particular flavor of error focuses on llime.a or liblime.a, like so: /Users[...]/lib/i386-debug/liblime.a(libfreetype.iphonesim.a), archive member 'libfreetype.iphonesim.a' with length 609744 is not mach-o or llvm bitcode for architecture i386

**Comment by jgranick** [comment]
Comment by __skylerparr__:

I was able to solve it by doing a clean build and it finally worked.  What library are you trying to force load?

**Comment by jgranick** [comment]
Comment by __SavedByZero__:

liblime.a.  Or should I be trying to force load all of them?  Also, when you built it from scratch, did you create a project in xcode first, then put the haxe/openfl files into that project folder, or copy the haxe/openfl files into a new folder and then type -lime build ios -debug first?

**Comment by jgranick** [comment]
Comment by __skylerparr__:

No, you should only force load the library you're intending to use. 
For example, if I'm trying to use the libhypfacebook.a ndll, and the error that I'm getting (which is sort of a red herring) is

```
  archive member 'libfreetype.iphonesim.a' with length 594288 is not mach-o or llvm bitcode for     architecture i386 
```

You want to force load the library libhypfacebook.a.
Are you making changes to the liblime and trying to compile with your own version of liblime?

**Comment by jgranick** [comment]
Comment by __SavedByZero__:

No, not touching liblime.  I tried forcing libhypfacebook.a and I still get the same error I pasted.  Cleaning the build and rebuilding didn't help either.  Hmm.

**Comment by jgranick** [comment]
Comment by __skylerparr__:

Have you tried building libhypfacebook and then force loading?

**Comment by jgranick** [comment]
Comment by __SavedByZero__:

You mean go into the HypFacebook project folder (instead of my project which is including it) and type -lime build ios, or is there some way to build individual libraries in my project without building the project?  Very confused here, sorry.  Building HypFacebook for Android was much easier than this.

**Comment by jgranick** [comment]
Comment by __skylerparr__:

No, in the hypfacebook source there is a directory called "project" and
there's a bash script called iphone.sh.  Go ahead and run that script.
 Additionally, there are some missing updates for Lime, so you can use my
fork (https://github.com/skylerparr/HypFacebook).  It also has updates to
the include file to force the library compilation.

Hope this helps.

**Comment by jgranick** [comment]
Comment by __SavedByZero__:

I updated my Hypfacebook folder with that branch, ran the iPhone.sh (sh
iPhone.sh), but unfortunately I still get the same error when I try to
build my project.  I'm wondering if something is wrong with my "other
linker flags" section here:

-lstd
-lregexp
-lzlib
-lcurl_ssl
-lpng
-ljpeg
-logg
-lvorbis
-lsqlite3
-lApplicationMain
-stdlib=libc++
-ObjC
-fobjc-arc
-force_load /Users/errataetal/FacebookRelay/bin/ios/FacebookRelay/lib/i386-debug/libhypfacebook.a
-llime

On Tue, Mar 4, 2014 at 3:38 PM, skylerparr notifications@github.com wrote:

> No, in the hypfacebook source there is a directory called "project" and
> there's a bash script called iphone.sh. Go ahead and run that script.
> Additionally, there are some missing updates for Lime, so you can use my
> fork (https://github.com/skylerparr/HypFacebook). It also has updates to
> the include file to force the library compilation.
> 
> Hope this helps.
> 
> On Tue, Mar 4, 2014 at 12:33 PM, Michael Greenhut
> notifications@github.comwrote:
> 
> > You mean go into the HypFacebook project folder (instead of my project
> > which is including it) and type -lime build ios, or is there some way to
> > build individual libraries in my project without building the project?
> > Very
> > confused here, sorry. Building HypFacebook for Android was much easier
> > than
> > this.
> > 
> > ## 
> > 
> > Reply to this email directly or view it on GitHub<
> > https://github.com/openfl/openfl-native/issues/27#issuecomment-36670619>
> > 
> > .
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/openfl/openfl-native/issues/27#issuecomment-36671185
> .

## 

---

Michael Greenhut
Writer, Game Developer, Clarion workshop graduate

**Comment by jgranick** [comment]
Comment by __skylerparr__:

try removing the -ObjC flag

**Comment by jgranick** [comment]
Comment by __SavedByZero__:

The good news is I don't see that error anymore.  The bad news is I see 19 others. Woot!  They all look like variants of this: 
Undefined symbols for architecture i386:
  "_FT_Bitmap_Embolden", referenced from:
      lime::FreeTypeFont::LoadBitmap(int) in liblime.a(FreeType.o)

How many files should I have in Build phases/compile sources?  Right now I just have Main.mm (from my project).  Putting HypFacebook.mm and HypFacebook.h seemed to just add new errors.

**Comment by jgranick** [comment]
Comment by __skylerparr__:

You should only have the Main.mm

Can you verify in XCode that the facebook.framework is listed?

**Comment by jgranick** [comment]
Comment by __SavedByZero__:

FacebookSDK.framework is included, yes.  It's also in the "Link Binary with libraries".  
I have: 
FacebookSDK
UIKit
OpenAL
OpenGLES
QuartzCore
CoreAudio
CoreGraphics
AudioToolbox
AVFoundation
Foundation
Security
Social
Accounts

**Comment by jgranick** [comment]
Comment by __SavedByZero__:

Finally got it.  That last hitch was due to an issue in HypFacebook.mm --
there was a line that said NMEAppDelegate which had to be changed to
LimeAppDelegate, then my project had to be updated from lime in Terminal
(which made me have to reinstall the facebook sdk and re-modify the .plist,
but that's minor compared to resolving this issue).

On Tue, Mar 4, 2014 at 4:47 PM, skylerparr notifications@github.com wrote:

> You should only have the Main.mm
> 
> Can you verify in XCode that the facebook.framework is listed?
> 
> On Tue, Mar 4, 2014 at 1:40 PM, Michael Greenhut
> notifications@github.comwrote:
> 
> > The good news is I don't see that error anymore. The bad news is I see 19
> > others. Woot! They all look like variants of this:
> > Undefined symbols for architecture i386:
> > "_FT_Bitmap_Embolden", referenced from:
> > lime::FreeTypeFont::LoadBitmap(int) in liblime.a(FreeType.o)
> > 
> > How many files should I have in Build phases/compile sources? Right now I
> > just have Main.mm (from my project). Putting HypFacebook.m and
> > HypFacebook.h seemed to just add new errors.
> > 
> > ## 
> > 
> > Reply to this email directly or view it on GitHub<
> > https://github.com/openfl/openfl-native/issues/27#issuecomment-36681529>
> > 
> > .
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/openfl/openfl-native/issues/27#issuecomment-36682325
> .

## 

---

Michael Greenhut
Writer, Game Developer, Clarion workshop graduate

**Comment by jgranick** [comment]
Comment by __joshua-jandyco__:

This is resolved with openfl/lime-tools#81 and openfl/lime#157 which force_load all thirdparty libraries.

**Comment by jgranick** [comment]
Comment by __louisbl__:

:+1:
